group-coordinator/group-coordinator-api/src/main/java/org/apache/kafka/coordinator/group/api/assignor/GroupSpec.java

```diff
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.kafka.coordinator.group.assignor;
+package org.apache.kafka.coordinator.group.api.assignor;
 
 import org.apache.kafka.common.Uuid;
```

Comment:
I noticed that not all interfaces have `@InterfaceStability.Unstable`. Could you share the context to me?

Reply:
For another, `GroupSpec#MemberSubscriptionSpec` return the interface but `GroupSpec#memberAssignment` return a map struct. If it is public API now, maybe return `MemberAssignment` is more flexible if we want to enrich it in the future?

Reply:
> I noticed that not all interfaces have @InterfaceStability.Unstable. Could you share the context to me?

Fixed. Thanks.

> For another, GroupSpec#MemberSubscriptionSpec return the interface but GroupSpec#memberAssignment return a map struct. If it is public API now, maybe return MemberAssignment is more flexible if we want to enrich it in the future?

Good question. Reusing `MemberAssignment` is not ideal because we want to have the ability to pass internal objects. In this case, it could be backed by the `Assignment`. So, we could think of using an interface though.

On thing to consider is whether the `MemberAssignment` returned by the assignor would always be the same as the one provided to the assignor. The actual assignment (the map) is for sure. What do you think?

Reply:
>  So, we could think of using an interface though.

+1

> On thing to consider is whether the MemberAssignment returned by the assignor would always be the same as the one provided to the assignor. The actual assignment (the map) is for sure. What do you think?

not sure whether I have got the point. The member assignment in `GroupAssignment` returned from `PartitionAssignor#assign` is the new assignment. Hence, it should be different to assignment in `GroupSpec`.

Reply:
> You suggested to use MemberAssignment in your previous comment. MemberAssignment is already used in GroupAssignment returned by PartitionAssignor#assign. I am not sure if you meant to reuse the same MemberAssignment in both places.

ummm, I guess I did not notice MemberAssignment is used by `GroupAssignment` before :_

Anyway, having a interface to replace `Map` struct is good way to me. Or we can keep current version if the tag "unstable" give the room to modify those public stuff :)

Reply:
I am playing with this. I will suggest something a bit later today.

Reply:
@chia7712 I gave it a go. You can check the last commit to see how it looks like. I also moved some classes between internal packages while here.